draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1216 @@
+    format which allows machines to parse them more easily.
+
+  JSON
+    JSON stands for JavaScript Object Notation, which is a way to format data so
+    that it can be transmitted from one place to another, most commonly between
+    a server and a Web application.
+
+  stdout
+    Stdout, also known as standard output, is the default file descriptor
+    where a process can write output.
+
+Message Types
+-------------
+
+In relation to Celery :term:`Command messages <Command Message>`
+are the messages we publish whenever we delay a task.
+:term:`Document messages <Document Message>` are the messages we get as a result.
+
+.. code-block:: pycon
+
+  >>> from celery import task
+  >>> @task
+  ... def add(a, b):
+  ...   return a + b
+  >>> result = add.delay(1, 2)  # Publish a command message
+  >>> result.get()  # Consume a Document message
+  3
+
+:term:`Event messages <Event Message>` are a new concept for Celery.
+They describe that a domain event occurred.
+Multiple tasks can be subscribed to an event.
+
+.. code-block:: pycon
+
+  >>> from uuid import UUID
+  >>> from celery import task, event
+  >>> from myapp.models import User, AccountManager
+  >>> @task
+  ... def send_welcome_email(user_id, email):
+  ...   send_email(email=email, contents="hello, welcome", subject="welcome") # Send a welcome email to the user...
+  ...   User.objects.filter(pk=user_id).update(welcome_email_sent=True)
+  >>> @task
+  ... def notify_account_manager(user_id, email):
+  ...   account_manager = AccountManager.objects.assign_account_manager(user_id)
+  ...   send_email(email=account_manager.email, contents="you have a new user to attend to", subject="Alert") # Send an email to the account manager...
+  >>> @event
+  ... class UserRegistered:
+  ...   user_id: UUID
+  ...   email: str
+  >>> UserRegistered.subscribe(send_welcome_email)
+  >>> UserRegistered.subscribe(notify_account_manager)
+  >>> UserRegistered.delay(user_id=1, email='foo@bar.com')  # Calls both send_welcome_email and notify_account_manager with the provided arguments.
+
+These architectural building blocks will aid us in creating a better messaging
+system. To encourage `ubiquitous language`_, we will be using them in this
+document and in Celery 5's codebase as well.
+
+Observability
+-------------
+
+One of Celery 5's goals is to be :term:`observable <Observability>`.
+
+Each Celery component will record statistics, provide trace points for
+application monitoring tools and distributed tracing tools and emit log messages
+when appropriate.
+
+Metrics
++++++++
+
+Celery stores and publishes metrics which allows our users to debug their
+applications more easily and spot problems.
+
+By default each worker will publish the metrics to a dedicated queue.
+
+Other methods such as publishing them to StatsD is also possible using the
+provided extension point.
+
+Trace Points
+++++++++++++
+
+Celery provides trace points for application monitoring tools and distributed
```

Comment:
Perhaps [OpenTracing](https://opentracing.io/guides/python/) can provide a good guide, on how tracing could be integrated by users, and we could add it for reference.

Reply:
I agree. Uber has an open tracing framework in Python.
I'll look into it.